03-events-forms-refs-life-cycle/forms-and-refs/part1/src/AddDivForm.js

```diff
@@ -0,0 +1,103 @@
+import React, { Component } from "react";
+
+class AddDivForm extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      width: "",
+      height: "",
+      backgroundColor: ""
+    };
+    this.handleChange = this.handleChange.bind(this);
+  }
+  handleChange(e) {
+    if (e.target.id === "backgroundColor")
+      this.setState({ [e.target.id]: e.target.value.slice(0, 6) });
+    else this.setState({ [e.target.id]: e.target.value.slice(0, 3) });
+  }
+
+  handleSubmit(e) {
+    e.preventDefault();
+    this.props.fnAdd({
+      width: this.state.width + "px",
+      height: this.state.height + "px",
+      backgroundColor: "#" + this.state.backgroundColor
+    });
+    this.setState({
+      width: "",
+      height: "",
+      backgroundColor: ""
+    });
+  }
+
+  render() {
+    let error = this.validate();
+    return (
+      <form>
+        {error || <h3> Yay! Now Submit!!! </h3>}
+        <label htmlFor="width">Width:</label>
+        <input
+          name="width"
+          id="width"
+          placeholder="Put width in px"
+          value={this.state.width}
+          onChange={this.handleChange}
+        />
+        <label htmlFor="height">Height:</label>
+        <input
+          name="height"
+          id="height"
+          placeholder="Put height in px"
+          value={this.state.height}
+          onChange={this.handleChange}
+        />
+        <label htmlFor="backgroundColor">Color:</label>
+        <input
+          name="backgroundColor"
+          id="backgroundColor"
+          placeholder="Put backgroundColor in hex color code"
+          value={this.state.backgroundColor}
+          onChange={this.handleChange}
+        />
+        {error ? (
+          <button onClick={e => e.preventDefault()}>Can't submit yet</button>
+        ) : (
+          <button onClick={this.handleSubmit.bind(this)}>Add new div</button>
+        )}
+      </form>
+    );
+  }
+  //Is this the best place to put this
+  validate() {
```

Comment:
it's fine as a class method, although you could make it a pure function outside of the class so it's more testable, then simply call it with `this.state` args.